File: artifacts/api-server/src/routes/stats.ts
import { Router, type IRouter } from "express";
import { eq, count, and } from "drizzle-orm";
import { db, usersTable, districtsTable, requestsTable, tagsTable, requestTagsTable } from "@workspace/db";
import { GetDistrictStatsParams } from "@workspace/api-zod";

const router: IRouter = Router();

async function getTopTags(limit = 5) {
  const tags = await db.select().from(tagsTable).orderBy(tagsTable.name);
  const tagsWithCounts = await Promise.all(tags.map(async (tag) => {
    const [result] = await db.select({ count: count() }).from(requestTagsTable).where(eq(requestTagsTable.tagId, tag.id));
    return { id: tag.id, name: tag.name, color: tag.color, requestCount: Number(result?.count ?? 0) };
  }));
  return tagsWithCounts.sort((a, b) => b.requestCount - a.requestCount).slice(0, limit);
}

router.get("/stats/overview", async (_req, res): Promise<void> => {
  const [totalUsersResult] = await db.select({ count: count() }).from(usersTable);
  const [totalMentorsResult] = await db.select({ count: count() }).from(usersTable).where(eq(usersTable.role, "mentor"));
  const [totalMenteesResult] = await db.select({ count: count() }).from(usersTable).where(eq(usersTable.role, "mentee"));
  const [totalDistrictsResult] = await db.select({ count: count() }).from(districtsTable);
  const [openRequestsResult] = await db.select({ count: count() }).from(requestsTable).where(eq(requestsTable.status, "open"));
  const [matchedRequestsResult] = await db.select({ count: count() }).from(requestsTable).where(eq(requestsTable.status, "matched"));

  const topTags = await getTopTags(5);

  res.json({
    totalUsers: Number(totalUsersResult?.count ?? 0),
    totalMentors: Number(totalMentorsResult?.count ?? 0),
    totalMentees: Number(totalMenteesResult?.count ?? 0),
    totalDistricts: Number(totalDistrictsResult?.count ?? 0),
    openRequests: Number(openRequestsResult?.count ?? 0),
    successfulMatches: Number(matchedRequestsResult?.count ?? 0),
    topTags,
  });
});

router.get("/stats/district/:id", async (req, res): Promise<void> => {
  const params = GetDistrictStatsParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }

  const [district] = await db.select().from(districtsTable).where(eq(districtsTable.id, params.data.id));
  if (!district) {
    res.status(404).json({ error: "District not found" });
    return;
  }

  const [memberCountResult] = await db.select({ count: count() }).from(usersTable).where(eq(usersTable.districtId, params.data.id));
  const [mentorCountResult] = await db.select({ count: count() }).from(usersTable).where(
    and(eq(usersTable.districtId, params.data.id), eq(usersTable.role, "mentor"))
  );
  const [menteeCountResult] = await db.select({ count: count() }).from(usersTable).where(
    and(eq(usersTable.districtId, params.data.id), eq(usersTable.role, "mentee"))
  );
  const [openRequestsResult] = await db.select({ count: count() }).from(requestsTable).where(
    and(eq(requestsTable.districtId, params.data.id), eq(requestsTable.status, "open"))
  );
  const [matchedRequestsResult] = await db.select({ count: count() }).from(requestsTable).where(
    and(eq(requestsTable.districtId, params.data.id), eq(requestsTable.status, "matched"))
  );

  const topTags = await getTopTags(5);

  res.json({
    districtId: district.id,
    districtName: district.name,
    memberCount: Number(memberCountResult?.count ?? 0),
    mentorCount: Number(mentorCountResult?.count ?? 0),
    menteeCount: Number(menteeCountResult?.count ?? 0),
    openRequests: Number(openRequestsResult?.count ?? 0),
    matchedRequests: Number(matchedRequestsResult?.count ?? 0),
    topTags,
  });
});

export default router;

File: artifacts/api-server/python/routers/stats.py
from fastapi import APIRouter, HTTPException
from db import db

router = APIRouter()

def get_top_tags(cur, limit=5):
    cur.execute("SELECT * FROM tags ORDER BY name")
    tags = cur.fetchall()
    result = []
    for tag in tags:
        cur.execute("SELECT COUNT(*) as cnt FROM request_tags WHERE tag_id = %s", (tag["id"],))
        count = cur.fetchone()["cnt"]
        result.append({"id": tag["id"], "name": tag["name"], "color": tag["color"], "requestCount": count})
    result.sort(key=lambda t: t["requestCount"], reverse=True)
    return result[:limit]

@router.get("/stats/overview")
def stats_overview():
    with db() as conn:
        cur = conn.cursor()
        cur.execute("SELECT COUNT(*) as cnt FROM users")
        total_users = cur.fetchone()["cnt"]
        cur.execute("SELECT COUNT(*) as cnt FROM users WHERE role = 'mentor'")
        total_mentors = cur.fetchone()["cnt"]
        cur.execute("SELECT COUNT(*) as cnt FROM users WHERE role = 'mentee'")
        total_mentees = cur.fetchone()["cnt"]
        cur.execute("SELECT COUNT(*) as cnt FROM districts")
        total_districts = cur.fetchone()["cnt"]
        cur.execute("SELECT COUNT(*) as cnt FROM requests WHERE status = 'open'")
        open_requests = cur.fetchone()["cnt"]
        cur.execute("SELECT COUNT(*) as cnt FROM requests WHERE status = 'matched'")
        successful_matches = cur.fetchone()["cnt"]
        top_tags = get_top_tags(cur)

    return {
        "totalUsers": total_users,
        "totalMentors": total_mentors,
        "totalMentees": total_mentees,
        "totalDistricts": total_districts,
        "openRequests": open_requests,
        "successfulMatches": successful_matches,
        "topTags": top_tags,
    }

@router.get("/stats/district/{district_id}")
def district_stats(district_id: int):
    with db() as conn:
        cur = conn.cursor()
        cur.execute("SELECT * FROM districts WHERE id = %s", (district_id,))
        district = cur.fetchone()
        if not district:
            raise HTTPException(status_code=404, detail="District not found")

        cur.execute("SELECT COUNT(*) as cnt FROM users WHERE district_id = %s", (district_id,))
        member_count = cur.fetchone()["cnt"]
        cur.execute("SELECT COUNT(*) as cnt FROM users WHERE district_id = %s AND role = 'mentor'", (district_id,))
        mentor_count = cur.fetchone()["cnt"]
        cur.execute("SELECT COUNT(*) as cnt FROM users WHERE district_id = %s AND role = 'mentee'", (district_id,))
        mentee_count = cur.fetchone()["cnt"]
        cur.execute("SELECT COUNT(*) as cnt FROM requests WHERE district_id = %s AND status = 'open'", (district_id,))
        open_requests = cur.fetchone()["cnt"]
        cur.execute("SELECT COUNT(*) as cnt FROM requests WHERE district_id = %s AND status = 'matched'", (district_id,))
        matched_requests = cur.fetchone()["cnt"]
        top_tags = get_top_tags(cur)

    return {
        "districtId": district["id"],
        "districtName": district["name"],
        "memberCount": member_count,
        "mentorCount": mentor_count,
        "menteeCount": mentee_count,
        "openRequests": open_requests,
        "matchedRequests": matched_requests,
        "topTags": top_tags,
    }
